File: jbake-core/src/main/java/org/jbake/app/configuration/Property.java
package org.jbake.app.configuration;

import java.util.Objects;

public class Property implements Comparable {

    private final String key;
    private final String description;
    private final Group group;

    public Property(String key, String description) {
        this(key, description, Group.DEFAULT);
    }

    public Property(String key, String description, Group group) {
        this.key = key;
        this.description = description;
        this.group = group;
    }

    public String getKey() {
        return key;
    }

    public String getDescription() {
        return description;
    }

    public Group getGroup() {
        return group;
    }

    @Override
    public String toString() {
        return getKey();
    }

    @Override
    public boolean equals(Object o) {
        if (this == o) {
            return true;
        }
        if (o == null || getClass() != o.getClass()) {
            return false;
        }
        Property property = (Property) o;
        return Objects.equals(key, property.key) &&
            Objects.equals(description, property.description) &&
            group == property.group;
    }

    @Override
    public int hashCode() {
        return Objects.hash(key, description, group);
    }

    @Override
    public int compareTo(Object o) {
        Property other = (Property) o;
        int result = this.getGroup().compareTo(other.getGroup());

        if (result == 0) {
            result = this.getKey().compareTo(other.getKey());
        }
        return result;
    }

    public enum Group {
        DEFAULT, CUSTOM
    }



}

File: jbake-core/src/main/java/org/jbake/app/configuration/Property.kt
package org.jbake.app.configuration

import java.util.*

class Property @JvmOverloads constructor(
    @JvmField val key: String,
    @JvmField val description: String?,
    @JvmField val group: Group = Group.DEFAULT
) : Comparable<Property> {

    override fun toString() = this.key

    override fun equals(other: Any?): Boolean {
        if (this === other) return true
        if (other == null || javaClass != other.javaClass) return false

        val property = other as Property
        return key == property.key &&
                description == property.description && group == property.group
    }

    override fun hashCode(): Int {
        return Objects.hash(key, description, group)
    }

    override fun compareTo(other: Property): Int {
        var result = this.group.compareTo(other.group)

        if (result == 0) {
            result = this.key.compareTo(other.key)
        }
        return result
    }

    enum class Group {
        DEFAULT, CUSTOM
    }
}
